# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# SPI Host use the defauly ByteOrder=1 which means it will shifts out the LSB byte first.

diff --git a/src/spi_host/model/spiflash.sv b/src/spi_host/model/spiflash.sv
index 52eb0a0..705c333 100644
--- a/src/spi_host/model/spiflash.sv
+++ b/src/spi_host/model/spiflash.sv
@@ -356,6 +356,12 @@ program spiflash #(
       default: mode = Addr3B;
     endcase
 
+    // The SPI host talking to this Flash sends LSBs first
+    for(int i = 0; i < 3; i++) begin
+      get_byte(imode, spi_byte);
+      addr[8*i+:8] = spi_byte;
+    end
+
     if (mode == Addr4B) begin
       // Get upper byte
       get_byte(imode, spi_byte);
@@ -363,12 +369,6 @@ program spiflash #(
     end else begin
       addr[31:24] = 8'h 0;
     end
-
-    for(int i = 2 ; i >= 0 ; i--) begin
-      get_byte(imode, spi_byte);
-      addr[8*i+:8] = spi_byte;
-    end
-
   endtask : get_address
 
   task automatic return_byte(spiflash_byte_t data, io_mode_e io_mode);
